Refactor

Yadunund opened the pull request:
Warning: These changes break API

## Refactor
### Summary of changes:
1. `battery_system`, `motion_sink`, `ambient_sink` and `planner` bundled into new `rmf_task::agv::Parameters` class
2. Updated signature for `rmf_task::agv::TaskPlanner::Configuration`. 
```   
Configuration(
      Parameters parameters,
      Constraints constraints,
      ConstCostCalculatorPtr cost_calculator,
      bool drain_battery);
 ```
3. `TaskPlanner::greedy_plan()` and `TaskPlanner::optimal_plan()` no longer need `std::vector<Constraints>` to be passed in. The `TaskPlanner` always assumed that the battery characteristics and nav graphs of the agents being planned for are same. In facts the `Requests` are generated with this assumption. But users were able to pass a `Constraint` for each `State` implying that agents may have unique characteristics. This was misleading. Now, by passing in a single `Constraints` into `TaskPlanner::Configuration`, it is clearer that these constraints apply to all agents.
4. A `recharge_soc` parameter is added to `rmf_task::agv::Constraints` to allow users to specify what battery level an agent should be recharged up to. Previously, this was always `1.0`. Fix https://github.com/open-rmf/rmf_task/issues/11
5. Previously, a `drain_battery` boolean was passed into the factory for any `Request`. This tied up battery drain with the request itself. Thus in downstream fleet adapter packages, it was not possible to simply replan the assignments without factoring battery drain unless all the requests were regenerated. Now, the `drain_battery` parameter is specified in `TaskPlanner::Configuration` and the `Request::Description::estimate_finish()` function takes in a `drain_battery` boolean. Thus, the same set of requests can be planned for with/without battery drain by appropriately initializing the `TaskPlanner`. Note: A user will need to re-initialize the `TaskPlanner` when they switch `drain_battery` values as the internal cache will need to be reset. The API enforces this.

mxgrey commented on rmf_task/include/rmf_task/Request.hpp:
```diff
@@ -30,10 +30,11 @@
 {
 public:
+  /// Implement this for new type of requests.
   class Description
```
While we're refactoring the API, I have a suggestion that would add another layer of complexity to the API, but I think it would help make it more sound, consistent, and stable.

Right now there's a risk that a task description won't match up correctly with the `Configuration` given to the task planner. It's possible that the `motion_sink`, `ambient_sink`, `planner`, etc could get jumbled up where different values are given to the task requests than to the task planner.

To guarantee consistency, we could change the API into this:

```C++
class Request
{
public:

  class Model
  {
  public:
    virtual std::optional<Estimate> estimate_finish(
      const agv::State& initial_state,
      const agv::Constraints& task_planning_constraints,
      const std::shared_ptr<EstimateCache> estimate_cache) const = 0;

    virtual rmf_traffic::Duration invariant_duration() const = 0;

    virtual ~Model();
  };

  class Description
  {
  public:
    virtual std::shared_ptr<Model> make_model(
      rmf_traffic::Time earliest_start_time,
      const Parameters& parameters) const = 0;

    virtual ~Description();
  };

  /* ... The rest of Request remains unchanged ... */
};
```

When the task planner takes in its vector of `requests`, it would then pass the `parameters` from its `configuration` into the `Description::make_model(~)` of each request to get the `Request::Model` instance that it can then use to calculate the estimates.

The APIs for producing tasks would be able to simplify like this:

```C++
class Clean
{
public:

  // Note: the Model class definition for Clean doesn't necessarily
  // need to be exposed to the public API.
  class Model;

  class Description : public Request::Description
  {
  public:
    static DescriptionPtr make(
      std::size_t start_waypoint,
      std::size_t end_waypoint,
      const rmf_traffic::Trajectory& cleaning_path);

    std::shared_ptr<Request::Model> make_model(
      rmf_traffic::Time earliest_start_time,
      const Parameters& parameters) const final;

    std::size_t start_waypoint() const;
    std::size_t end_waypoint() const;
    rmf_traffic::agv::Planner::Start location_after_clean(
      rmf_traffic::agv::Planner::Start start) const;

    class Implementation;
  private:
    Description();
    rmf_utils::impl_ptr<Implementation> _pimpl;
  };

  static ConstRequestPtr make(
    std::size_t start_waypoint,
    std::size_t end_waypoint,
    const rmf_traffic::Trajectory& cleaning_path,
    const std::string& id,
    rmf_traffic::Time earliest_start_time,
    ConstPriorityPtr priority = nullptr);
};
```

This should help satisfy the single-source-of-truth principle and separate the concerns of the classes. `Description` would only be concerned with the generic description of the type of task, unconcerned with the properties of who/what is performing the task, while `Model` would be the tool used by the planner for actually calculating the outcome of performing the task in specific conditions for a specific robot.

Changing the API like this would also help in the future if we decide to offer a task planner API that supports planning across multiple different types of robot platforms simultaneously, because now the `Task::Description` instances are no longer tied to specific robot parameters.

mxgrey reviewed (changes requested):
This is definitely a good direction for the API to move towards.

As we approach wanting a stable long-term release for `rmf_task` it's becoming increasingly important for us to use strategies that keep the API and ABI stable. To that end, I've recommended some additional refactoring that I believe will help to really solidify that API/ABI stability, even if we decide to make big sweeping implementation changes in the future.

The additional refactoring I'm recommended may involve some heavy lifting, but I think it'll be worth it for the future-proofing.

Yadunund commented:
Thanks for the guidance @mxgrey. I've refactored as per your suggestion.

mxgrey reviewed (changes requested):
Thanks for getting this finished so quickly! I think this is a huge improvement to the API, and you did a great job implementing it.

I just have some pretty minor nitpick suggestions remaining. There's just one that I'd like to emphasize, which is the `request_models` map. I think that approach is adding some unnecessary map lookup overhead that we can avoid pretty easily with a slight change to `PendingTask`.

Yadunund commented on rmf_task/include/rmf_task/Request.hpp:
```diff
@@ -20,21 +20,23 @@
+  class Model
   {
   public:
 
```
Done https://github.com/open-rmf/rmf_task/pull/15/commits/d08ec08a0b560b52c5f15b9d77c0a71e467f7b6b

Yadunund commented on rmf_task/src/rmf_task/requests/ChargeBattery.cpp:
```diff
@@ -125,6 +118,10 @@ ChargeBatteryDescription::estimate_finish(
+  const bool drain_battery = task_planning_constraints.drain_battery();
+  const auto planner = _parameters.planner();
+  const auto motion_sink = _parameters.motion_sink();
+  const auto device_sink = _parameters.ambient_sink();
```
Done https://github.com/open-rmf/rmf_task/pull/15/commits/15e20ac2d39a6d95d6ac9ed8c467769cfb9b0c14

Yadunund commented on rmf_task/src/rmf_task/agv/internal_task_planning.hpp:
```diff
@@ -216,9 +170,10 @@ struct Node
       double earliest_start_time = rmf_traffic::time::to_seconds(
         u.second.request->earliest_start_time().time_since_epoch());
+      const auto invariant_duration =
+        request_models[u.second.request->id()]->invariant_duration();
```
That's a much better way! Done.
https://github.com/open-rmf/rmf_task/pull/15/commits/5639b9f33cac87bac8c3b0d68363211801ae1208

Yadunund commented on rmf_task/include/rmf_task/requests/Clean.hpp:
```diff
@@ -28,70 +29,51 @@
-    std::shared_ptr<const rmf_traffic::agv::Planner> planner,
-    rmf_traffic::Time start_time,
-    bool drain_battery = true,
+    rmf_traffic::Time earliest_start_time,
```
Done https://github.com/open-rmf/rmf_task/pull/15/commits/be968cb913ac58deed2d007a26ecaac70cd81bf6

Yadunund commented on rmf_task/include/rmf_task/agv/Parameters.hpp:
```diff
@@ -0,0 +1,107 @@
+    rmf_battery::ConstDevicePowerSinkPtr ambient_sink);
+
+  /// Get the tool device sink
+  const rmf_battery::ConstDevicePowerSinkPtr& tool_sink() const;
```
Agreed, we need a better way to manage these devices.

Yadunund commented on rmf_task/include/rmf_task/requests/ChargeBattery.hpp:
```diff
@@ -27,64 +28,46 @@
 
-  rmf_traffic::Duration invariant_duration() const final;
+    static DescriptionPtr make(
+      double max_charge_soc = 1.0);
```
Good point. Removed `max_charge_soc` here. The `recharge_soc` will not be available via the `parameters` as it is in `Constraints`. But since we need this value only inside `estimate_finish`, we can get it from `task_planning_constraints` that is passed along.
https://github.com/open-rmf/rmf_task/pull/15/commits/4c36e84e6709e3d765fa357850b13eb1f62aa955

Yadunund commented:
Thanks for the feedback @mxgrey. I hope I have addressed all your points.

One thing I would like to point out is that with the refactor, the average time taken to solve the scenarios in the unit test goes down by 15-30%. Here are the test printouts [before](https://gist.github.com/Yadunund/7dd5ae43f0362f65199f15ad67ad478d) and [after](https://gist.github.com/Yadunund/a03a97e95f9361a056501014fec76a9f). The worst scenario takes ~160ms which is still not too bad. I'm guessing the increased time is from the initialization of the `Models` during each run which was previously "cached" into the `Requests` themselves. But the improved flexibility with the new API certainly outweighs the increased computation time.

mxgrey commented:
There isn't any reason we should have to accept a performance drop from the API changes. We probably just need one more implementation tweak.

Inside the planner code, define some struct like

```
struct RequestModel
{
  Request request;
  std::shared_ptr<Request::Model> model;
};
```

Initialize one of these at the start for each input request and then pass them around as a replacement for all the places that you used to pass a `Request` around.

Yadunund commented:
> Initialize one of these at the start for each input request and then pass them around as std::shared_ptr<const RequestModel> a replacement for all the places that you used to pass a std::shared_ptr<Request> around.

Sure, that would prevent Models from being initialized every time `make_initial_node()` is called but wouldn't the overall solve duration still be higher than earlier as we still need to initialize all the Models once? Of course if we measure performance by summing the durations to create the requests and then solve them, the performance will be the same with the change above. But in the fleet adapters for example where requests are generated once and are replanned every now and then, the time taken to produce a plan would be longer with the new API.

But yes, I shall implement the suggestion above.

mxgrey commented:
I see, I'm surprised that the one-time calculation of the invariants would cost so much. I guess that's from computing the energy drain while driving?

After looking more closely at the code, I'm not sure my suggestion about bundling the model with the request will accomplish anything since it looks like a `PendingTask` instance is only created once for each request.

I think we could deal with the replanning performance drop by beefing up the `TaskPlanner` or `TaskPlanner::Result` class to maintain a cache of the `Model` for each request in between runs, and have an API for us to add/remove/enable/disable requests before each planning attempt. But that can be saved for a followup PR.

mxgrey commented on rmf_task/src/rmf_task/requests/Clean.cpp:
```diff
@@ -23,92 +23,77 @@ namespace rmf_task {
+private:
+  rmf_traffic::Time _earliest_start_time;
+  agv::Parameters _parameters;
+  rmf_traffic::Trajectory& _cleaning_path;
```
We shouldn't use a raw mutable reference here. This is begging for trouble.

If the goal is to avoid duplicate `Trajectory` allocations, we should either have the `Model` take in a `std::shared_ptr<Clean::Description>` that gives it access to the trajectory, or we should have `Clean::Description` use a `std::shared_ptr<Trajectory>` that it can pass along to the `Model`.

Yadunund replied:
Realized `_cleaning_path` is unused so got rid of it. https://github.com/open-rmf/rmf_task/pull/15/commits/6490a3635f3d1c1e7e9339223de61eb29e18481b

mxgrey commented on rmf_task/src/rmf_task/requests/Loop.cpp:
```diff
@@ -21,64 +21,63 @@ namespace rmf_task {
+  agv::Parameters _parameters;
+  std::size_t _start_waypoint;
+  std::size_t _finish_waypoint;
+  std::size_t _num_loops;
```
Nitpick: This field is unused, so we should remove it.

mxgrey reviewed (changes requested):
Just two last small requests!

Yadunund commented:
> After looking more closely at the code, I'm not sure my suggestion about bundling the model with the request will accomplish anything since it looks like a PendingTask instance is only created once for each request.

Yeah I tried it out but there wasn't any significant improvement.

> I think we could deal with the replanning performance drop by beefing up the TaskPlanner or TaskPlanner::Result class to maintain a cache of the Model for each request in between runs, and have an API for us to add/remove/enable/disable requests before each planning attempt. But that can be saved for a followup PR.

Sounds good. I'll open a ticket for that.

Yadunund commented:
> I just noticed, I think there was a misunderstanding. I was saying we can change this whole if-statement to
> 
> ```
> if (initial_state.battery_soc() >= recharge_soc - 1e-3
>   && initial_state.waypoint() == initial_state.charging_waypoint())
> ```
> 
> The way it is right now, the first check in the or-statement is redundant with the second.

whoops gross oversight 🤦🏼‍♂️

mxgrey reviewed (approved):
Thanks for all the rapid improvements, this is looking great :+1:
